### PR TITLE
Feature/create category

### PR DESCRIPTION
**Pull request by maurosaravia**
## Pull request type

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

- Categories exist in the system and can be obtained
- Categories have a name, and can have a parent category. Categories with a parent are called subcategories
- A subcategory can't have children

## What is the new behavior?

- Categories can be added
- The system has clear error messages in case of an incorrect use

https://trello.com/c/Ed3nFyIG
https://trello.com/c/zNQhh19d

## Preview

N/A

### CODE REVIEW

File: src/categories/services/categories.service.ts [hunk 1]
```diff
@@ -34,4 +36,33 @@ export class CategoriesService {
...
+        parent = await this.categoriesRepo.findById(dto.parentCategoryId);
+      } catch (exception) {
+        if (exception instanceof EntityNotFoundError)
+          throw new BadRequestException('Parent category does not exist');
+        throw new InternalServerErrorException();
+      }
+
+      //The new parent category is a subcategory and can't have children
+      if (parent.parentCategoryId) {
+        throw new BadRequestException("Parent category can't be a subcategory");
```
Comment (slasia) [inline src/categories/services/categories.service.ts:54]: Youu can create a custom Exception class to log that information.  

for example: 


```javascript
export class ParentCategoryException extends HttpException {
  constructor(stacktrace: string) {
    Logger.error('Parent category can't be a subcategory. Stacktrace:', stacktrace);
    super('An error ocurred while saving the category. Please try again later.', HttpStatus.BAD_REQUEST);
  }
}
```

File: src/categories/services/categories.service.ts [hunk 2]
```diff
@@ -34,4 +36,33 @@ export class CategoriesService {
...
+    }
+  }
+
+  async create(dto: CategoryDTO): Promise<Category> {
+    await this.validateDTO(dto);
+    try {
+      const category = await this.categoriesRepo.createFromDto(dto);
+      return category;
+    } catch (exception) {
+      throw new InternalServerErrorException();
```
Comment (slasia) [inline src/categories/services/categories.service.ts:65]: Same last comment

File: src/categories/services/categories.service.ts (hunk 1 above)
Comment (CamilaNiebles) [reply src/categories/services/categories.service.ts:54]: Also, I guess that this validateDTO function can be a middleware. What do you think?

File: src/categories/services/categories.service.ts [hunk 3]
```diff
@@ -34,4 +36,33 @@ export class CategoriesService {
       throw new InternalServerErrorException();
     }
   }
+
+  async validateDTO(dto: CategoryDTO): Promise<void> {
```
Comment (EMazzaglia) [inline src/categories/services/categories.service.ts:40]: You can create your own [class-validator decorators](https://github.com/typestack/class-validator#custom-validation-decorators) to use on your dtos

File: src/categories/controllers/categories.controller.spec.ts [hunk 4]
```diff
@@ -34,4 +35,13 @@ describe('CategoriesController', () => {
       ...mockCategory,
     });
   });
+
+  it('should create a category', () => {
```
Comment (EMazzaglia) [inline src/categories/controllers/categories.controller.spec.ts:39]: There are some conventions here to follow: 

When writing unit tests you can adopt the naming convention of 

Grouping the Class that you are gonna tests inside a describe block 

And then write each unit tests with following the name 

SHOULD-WHEN

**Pull request merged**